nimble/transport/apollo3/src/apollo3_ble_hci.c

```diff
@@ -0,0 +1,354 @@
+        /* ui32ChipID0 bit 8-31 is test time during chip manufacturing */
+        g_ble_mac_address[4] = (device_info.ui32ChipID0 >> 8) & 0xFF;
+        g_ble_mac_address[5] = (device_info.ui32ChipID0 >> 16) & 0xFF;
+    }
+
+    NVIC_EnableIRQ(BLE_IRQn);
+
+    return 0;
+}
+
+/* Wake update helper function */
+static void
+apollo3_update_wake(void)
+{
+    AM_CRITICAL_BEGIN;
+
+    /* Set WAKE if there's something in the write queue, but not if SPISTATUS or IRQ is high. */
+    if ((BLEIFn(0)->BSTATUS_b.SPISTATUS == 0) && (BLEIF->BSTATUS_b.BLEIRQ == false)) {
+        am_hal_ble_wakeup_set(ble_handle, 1);
+
+        /* If we've set wakeup, but IRQ came up at the same time, we should just lower WAKE again. */
+        if (BLEIF->BSTATUS_b.BLEIRQ == true) {
+            am_hal_ble_wakeup_set(ble_handle, 0);
+        }
+    }
+
+    AM_CRITICAL_END;
+}
+
+/*
+ * Function used by the BLE stack to send HCI messages to the BLE controller.
+ * The payload is placed into a queue and the controller is turned on. When it is ready
+ * an interrupt will fire to handle sending a message
+ */
+static uint8_t
+apollo3_hci_write(uint8_t type, uint16_t len, uint8_t *data)
+{
+    uint8_t *write_ptr;
+    hci_drv_write_t write_buf;
+
+    /* comparison compensates for the type byte at index 0. */
+    if (len > (MYNEWT_VAL(BLE_HCI_DRV_MAX_TX_PACKET)-1)) {
+        return 0;
+    }
+
+    /* Set all of the fields in the hci write structure. */
+    write_buf.len = len + 1;
+
+    write_ptr = (uint8_t *) write_buf.data;
+
+    *write_ptr++ = type;
+
+    for (uint32_t i = 0; i < len; i++) {
+        write_ptr[i] = data[i];
+    }
+
+    /* Wake up the BLE controller. */
+    apollo3_update_wake();
+
+    /* Wait on SPI status before writing */
+    while (BLEIFn(0)->BSTATUS_b.SPISTATUS) {
+        os_time_delay(1);
+    }
+
+    am_hal_ble_blocking_hci_write(ble_handle, AM_HAL_BLE_RAW, write_buf.data, write_buf.len);
+
+    return 0;
+}
+
+static int
+apollo3_ble_hci_acl_tx(struct os_mbuf *om)
+{
+    struct os_mbuf *x;
+    int rc = 0;
+
+    x = om;
+    while (x) {
+        rc = apollo3_hci_write(HCI_H4_ACL, x->om_len, x->om_data);
+        if (rc < 0) {
+            break;
+        }
+        x = SLIST_NEXT(x, om_next);
+    }
+
+    os_mbuf_free_chain(om);
+
+    return (rc < 0) ? BLE_ERR_MEM_CAPACITY : 0;
+}
+
+static int
+apollo3_ble_hci_frame_cb(uint8_t pkt_type, void *data)
+{
+    int rc;
+
+    switch (pkt_type) {
+    case HCI_H4_ACL:
+        rc = ble_transport_to_hs_acl(data);
+        break;
+    case HCI_H4_EVT:
+        rc = ble_transport_to_hs_evt(data);
+        break;
+    default:
+        assert(0);
+        break;
+    }
+
+    return rc;
+}
+
+static void
+apollo3_ble_hci_init(void)
+{
+    SYSINIT_ASSERT_ACTIVE();
+
+    /* Enable interrupt to handle read based on BLECIRQ */
+    NVIC_SetVector(BLE_IRQn, (uint32_t)apollo3_hci_int);
+
+    /* Initial coldboot configuration */
+    apollo3_hci_radio_boot(1);
+}
+
+#if MYNEWT_VAL(BLE_HOST)
```

Comment:
remove - it will be possible to use apollo3 with blehci, i.e. uart to controller without NimBLE host